modules/scope3/rtd/module.go

```diff
@@ -0,0 +1,332 @@
+	// Parse the OpenRTB request
+	var bidRequest openrtb2.BidRequest
+	if err := json.Unmarshal(payload, &bidRequest); err != nil {
+		return hookstage.HookResult[hookstage.RawAuctionRequestPayload]{}, nil
+	}
+
+	// Call Scope3 API
+	segments, err := m.fetchScope3Segments(ctx, &bidRequest)
+	if err != nil {
+		// Log error but don't fail the auction
+		return hookstage.HookResult[hookstage.RawAuctionRequestPayload]{
+			AnalyticsTags: hookanalytics.Analytics{
+				Activities: []hookanalytics.Activity{{
+					Name:   "scope3_fetch",
+					Status: hookanalytics.ActivityStatusError,
+					Results: []hookanalytics.Result{{
+						Status: hookanalytics.ResultStatusError,
+						Values: map[string]interface{}{"error": err.Error()},
+					}},
+				}},
+			},
+		}, nil
+	}
+
+	// Store segments in module context
+	if segmentStore, ok := miCtx.ModuleContext["segments"].(*sync.Map); ok {
+		segmentStore.Store("segments", segments)
+	}
+
+	// Store segments for later use - no mutation needed at this stage
+	changeSet := hookstage.ChangeSet[hookstage.RawAuctionRequestPayload]{}
+
+	return hookstage.HookResult[hookstage.RawAuctionRequestPayload]{
+		ChangeSet: changeSet,
+		AnalyticsTags: hookanalytics.Analytics{
+			Activities: []hookanalytics.Activity{{
+				Name:   "scope3_fetch",
+				Status: hookanalytics.ActivityStatusSuccess,
+				Results: []hookanalytics.Result{{
+					Status: hookanalytics.ResultStatusModify,
+					Values: map[string]interface{}{
+						"segments": segments,
+						"count":    len(segments),
+					},
+				}},
+			}},
+		},
+	}, nil
+}
+
+// HandleProcessedAuctionHook adds targeting keys to the response
+func (m *Module) HandleProcessedAuctionHook(
+	ctx context.Context,
+	miCtx hookstage.ModuleInvocationContext,
+	payload hookstage.ProcessedAuctionRequestPayload,
+) (hookstage.HookResult[hookstage.ProcessedAuctionRequestPayload], error) {
+	// Retrieve segments from module context
+	var segments []string
+	if segmentStore, ok := miCtx.ModuleContext["segments"].(*sync.Map); ok {
+		if val, ok := segmentStore.Load("segments"); ok {
+			segments = val.([]string)
+		}
+	}
+
+	if len(segments) == 0 {
+		return hookstage.HookResult[hookstage.ProcessedAuctionRequestPayload]{}, nil
+	}
+
+	// Add targeting keys to the request
+	changeSet := hookstage.ChangeSet[hookstage.ProcessedAuctionRequestPayload]{}
+	changeSet.AddMutation(
+		func(payload hookstage.ProcessedAuctionRequestPayload) (hookstage.ProcessedAuctionRequestPayload, error) {
+			// Add Scope3 segments as targeting keys for GAM
+			// Format: "gmp_eligible,gmp_plus_eligible" for easy GAM key-value targeting
+			reqWrapper := payload.Request
+			if reqWrapper.BidRequest.Ext == nil {
+				reqWrapper.BidRequest.Ext = json.RawMessage("{}")
+			}
+
+			var extMap map[string]interface{}
+			if err := json.Unmarshal(reqWrapper.BidRequest.Ext, &extMap); err != nil {
+				extMap = make(map[string]interface{})
+			}
+
+			// Add targeting keys that will be available to GAM
```

Comment:
as a potential module user, i would strongly prefer these also came back in bid.meta so i could decie if i want to send them togam or somewhere else and also how i format them if i do send them to gam. This could always happen in addition or be some sort of toggle

Reply:
right but since we're not bidding that probably isn't the right spot? I changed it so by default it goes to `response.ext.scope3.segments` but if you add `add_to_targeting: true` it will add to the targeting list for GAM. does that make sense?

Reply:
I think that location works, but you raise a good point, we need your result back even on a non-bid event. @bsardo any thoughts on the best place?